internal/controllers/machine/machine_controller.go

```diff
@@ -215,15 +220,29 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		}
 	}()
 
-	// Reconcile labels.
-	if m.Labels == nil {
-		m.Labels = make(map[string]string)
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	// Note: Finalizers in general can only be added when the deletionTimestamp is not set.
+	if !controllerutil.ContainsFinalizer(m, clusterv1.MachineFinalizer) && m.ObjectMeta.DeletionTimestamp.IsZero() {
+		controllerutil.AddFinalizer(m, clusterv1.MachineFinalizer)
+		return ctrl.Result{}, nil
+	}
```

Comment:
Wondering if we should also add the finalizer earlier, directly after the pause thing 
(we should probably have utils for both that we then use everywhere)

Reply:
Nice idea, let's do this as a follow up